File: app/src/main/java/com/dubedivine/samples/data/model/Question.java
package com.dubedivine.samples.data.model;

import java.util.Date;
import java.util.List;

/**
 * Created by divine on 2017/09/10.
 */

public class Question {

    private  String id;  // protected because it has a setter in the Elastic question child class
    private   String title;
    private String body;
    private long votes;
    private List<Comment> comments;
    private List<Answer> answers;
    private List<Tag> tags;  //todo: bad it should be mapping!!
    private User user; // the user the
    private String type;
    private Media video;
    private List<Media> files; //this can be combined with video dwag
    private Date createdAt = new Date();



    public Question(String title, String body, long votes, List<Tag> tags, String type) {
        this.title = title;
        this.body = body;
        this.votes = votes;
        this.tags = tags;
        this.type = type;
    }

    //for video we will add another constructor which has video here

    public String getId() {
        return id;
    }

    public List<Tag> getTags() {
        return tags;
    }

    public String getTitle() {
        return title;
    }

    public String getBody() {
        return body;
    }

    public long getVotes() {
        return votes;
    }

    public List<Comment> getComments() {
        return comments;
    }

    public List<Answer> getAnswers() {
        return answers;
    }

    public Media getVideo() {
        return video;
    }

    public String getType() {
        return type;
    }

    public List<Media> getFiles() {
        return files;
    }

    public Date getCreatedAt() {
        return createdAt;
    }

    public void setCreatedAt(Date createdAt) {
        this.createdAt = createdAt;
    }

    public void setComments(List<Comment> comments) {
        this.comments = comments;
    }

    public void setAnswers(List<Answer> answers) {
        this.answers = answers;
    }

    public void setVideo(Media video) {
        this.video = video;
    }

    public void setFiles(List<Media> files) {
        this.files = files;
    }


    @Override
    public String toString() {
        return "Question{" +
                "id='" + id + '\'' +
                ", title='" + title + '\'' +
                ", body='" + body + '\'' +
                ", votes=" + votes +
                ", comments=" + comments +
                ", answers=" + answers +
                ", tags=" + tags +
                ", user=" + user +
                ", type='" + type + '\'' +
                ", video=" + video +
                ", files=" + files +
                ", createdAt=" + createdAt +
                '}';
    }

    public void setUser(User user) {
        this.user = user;
    }
    public User getUser() {
        return user;
    }
}

File: app/src/main/java/com/dubedivine/samples/data/model/Question.kt
package com.dubedivine.samples.data.model

import com.google.gson.annotations.SerializedName
import java.io.Serializable
import java.util.*
import kotlin.collections.ArrayList

/**
 * Created by divine on 2017/09/10.
 */

// todo: should not deserialize the the nested objects?

class Question(val title: String, val body: String, var votes: Long, val tags: List<Tag>  //todo: bad it should be mapping!!
               , val type: String) : Serializable {
    //for video we will add another constructor which has video here
    var id: String? = null  // protected because it has a setter in the Elastic question child class
    var comments: ArrayList<Comment>? = null
    @SerializedName("answers")
    var answers: ArrayList<Answer>? = null
    var user: User? = null // the user the
    var video: Media? = null
    var files: ArrayList<Media>? = null //this can be combined with video dwag
    var createdAt = Date()
    var answered: Boolean? = null


    override fun toString(): String {
        return "Question{" +
                "id='" + id + '\'' +
                ", title='" + title + '\'' +
                ", body='" + body + '\'' +
                ", votes=" + votes +
                ", comments=" + comments +
                ", answers=" + answers +
                ", tags=" + tags +
                ", user=" + user +
                ", type='" + type + '\'' +
                ", video=" + video +
                ", files=" + files +
                ", createdAt=" + createdAt +
                '}'
    }

    companion object {
        val TYPE_Q ="Q"
        val TYPE_P ="P"
    }
}
